controllers/user/password.js

```diff
@@ -0,0 +1,240 @@
+const xss = require('xss');
+const jwt = require('jsonwebtoken');
+const { validationResult } = require('express-validator/check');
+
+const models = require('../../models/index');
+const mail = require('../../modules/mail');
+const secrets = require('../../modules/secrets');
+const { userBasePath, userEndpoints, makeUserLink, renderUserError } = require('./utils');
+
+const requestResetForm = (req, res) => {
+    res.cacheControl = { maxAge: 0 };
+    res.render('user/resetpassword', {
+        mode: 'enterEmail',
+        makeUserLink: makeUserLink
+    });
+};
+
+// is this user in password update mode?
+const checkUserRequestedPasswordReset = (userId, callbackSuccess, callbackError) => {
+    models.Users
+        .findOne({
+            where: {
+                id: userId,
+                is_password_reset: true
+            }
+        })
+        .then(user => {
+            if (!user) {
+                // no user for this ID, or they already did this reset
+                return callbackError();
+            } else {
+                return callbackSuccess(user);
+            }
+        })
+        .catch(err => {
+            callbackError(err);
+        });
+};
+
+const changePasswordForm = (req, res) => {
+    res.cacheControl = { maxAge: 0 };
+    let token = req.query.token;
+    if (!token) {
+        return res.redirect(userBasePath + userEndpoints.login);
+    }
+
+    // a user has followed a reset link
+    jwt.verify(token, secrets['user.jwt.secret'], (err, decoded) => {
+        if (err) {
+            console.error('Password reset token expired', err);
+            req.flash('genericError', 'Your password reset period has expired - please try again');
+            return renderUserError(null, req, res, makeUserLink('resetpassword'));
+        } else {
+            if (decoded.data.reason === 'resetpassword') {
+                // now check if this user actually requested this
+                checkUserRequestedPasswordReset(
+                    decoded.data.userId,
+                    user => {
+                        // we can now show the reset password form
+                        return res.render('user/resetpassword', {
+                            mode: 'pickNewPassword',
+                            token: token,
+                            makeUserLink: makeUserLink
+                        });
+                    },
+                    error => {
+                        console.error('User attempted to reset a password for an non-resettable user', error || {});
+                        req.flash('genericError', 'Your password reset link was invalid - please try again');
+                        return renderUserError(null, req, res, makeUserLink('resetpassword'));
```

Comment:
👃  A little wary of functions that start with null as an argument…should this take an object instead?